library/Cube/IcingaDb/IcingaDbCube.php

```diff
@@ -0,0 +1,209 @@
+    protected $fullQuery;
+
+    /** @var boolean */
+    protected $isSliced;
+
+    /** @var array */
+    protected $groupBy;
+
+    /** @var array */
+    protected $columns;
+
+    abstract public function getObjectsFilter();
+
+    public function hasBeenSliced()
+    {
+        return $this->isSliced;
+    }
+
+    public function getRelation()
+    {
+        return $this->relation;
+    }
+
+    /**
+     * An IcingaDbCube must provide a list of all available columns
+     *
+     * This is a key/value array with the key being the fact name / column alias
+     * and
+     *
+     * @return array
+     */
+    abstract public function getAvailableFactColumns();
+
+    /**
+     * @return Query
+     */
+    abstract public function prepareInnerQuery();
+
+    /**
+     * Get our inner query
+     *
+     * Hint: mostly used to get rid of NULL values
+     *
+     * @return Query
+     */
+    public function innerQuery()
+    {
+        if ($this->innerQuery === null) {
+            $this->innerQuery = $this->prepareInnerQuery();
+        }
+
+        return $this->innerQuery;
+    }
+
+    /**
+     * Get our rollup query
+     *
+     * @return Select
+     */
+    protected function rollupQuery()
+    {
+        if ($this->rollupQuery === null) {
+            $this->rollupQuery = $this->prepareRollupQuery();
+        }
+
+        return $this->rollupQuery;
+    }
+
+    /**
+     * Add a specific named dimension
+     *
+     * @param string $name
+     * @return $this
+     */
+    public function addDimensionByName($name)
+    {
+        $this->addDimension($this->createDimension($name));
+
+        return $this;
+    }
+
+    /**
+     * We first prepare the queries and to finalize it later on
+     *
+     * This way dimensions can be added one by one, they will be allowed to
+     * optionally join additional tables or apply other modifications late
+     * in the process
+     *
+     * @return void
+     */
+    protected function finalizeInnerQuery()
+    {
+        $innerQuery = $this->prepareInnerQuery();
+        $resolver = $innerQuery->getResolver();
+        $dimensions = $this->listDimensions();
+        $sourceTable = $innerQuery->getModel()->getTableName();
+
+        $groupBy = [];
+        foreach ($resolver->resolveRelations($sourceTable . '.vars') as $relation) {
+            foreach ($dimensions as $name => $dimension) {
+                $quotedDimension = $this->getDb()->quoteIdentifier($name);
+                $this->relation = $relation;
+                $dimension->addToCube($this);
+
+                $columns[$quotedDimension] = 'f.' . $quotedDimension;
+                $groupBy[] = $quotedDimension;
+            }
+        }
+
+        foreach ($resolver->resolveRelations($sourceTable . '.vars') as $relation) {
+            foreach ($this->listSlices() as $name) {
+                $this->relation = $relation;
+                $dimension = $this->createDimension($name);
```

Comment:
Instead of having to call `createDimension` you could iterate over `$this->dimensions` above instead. It includes all dimensions no matter if shown or sliced. ($this->listDimensions() filters sliced ones out) That's also how the `DbCube` does it. With `$this->hasSlice()` you can check whether a filter is required or a column.